pontoon/api/schema.py

```diff
@@ -0,0 +1,120 @@
+        only_fields = (
+            'total_strings', 'approved_strings', 'translated_strings',
+            'fuzzy_strings', 'project', 'locale')
+
+
+class Project(DjangoObjectType, StringStats):
+    class Meta:
+        model = ProjectModel
+        only_fields = (
+            'name', 'slug', 'info', 'deadline', 'priority', 'contact',
+            'total_strings', 'approved_strings', 'translated_strings',
+            'fuzzy_strings')
+
+    localizations = graphene.List(ProjectLocale)
+
+    @graphene.resolve_only_args
+    def resolve_localizations(self):
+        return self.project_locale.all()
+
+
+class Locale(DjangoObjectType, StringStats):
+    class Meta:
+        model = LocaleModel
+        only_fields = (
+            'name', 'code', 'direction', 'script', 'population',
+            'total_strings', 'approved_strings', 'translated_strings',
+            'fuzzy_strings')
+
+    localizations = graphene.List(ProjectLocale)
+
+    @graphene.resolve_only_args
+    def resolve_localizations(self):
+        return self.project_locale.all()
+
+
+class Query(graphene.ObjectType):
+    debug = graphene.Field(DjangoDebug, name='__debug')
+
+    projects = graphene.List(Project)
+    project = graphene.Field(Project, slug=graphene.String())
+
+    locales = graphene.List(Project)
+    locale = graphene.Field(Locale, code=graphene.String())
+
+    def resolve_projects(self, args, context, info):
+        qs = ProjectModel.objects.all()
+        fields = get_fields(info)
+
+        if 'projects.localizations' in fields:
+            qs = qs.prefetch_related('project_locale__locale')
+
+        if 'projects.localizations.locale.localizations' in fields:
+            raise Exception('Cyclic queries are forbidden')
+
+        return qs
+
+    def resolve_project(self, args, context, info):
+        qs = ProjectModel.objects
+        fields = get_fields(info)
+
+        if 'project.localizations' in fields:
+            qs = qs.prefetch_related('project_locale__locale')
+
+        if 'project.localizations.locale.localizations' in fields:
+            raise Exception('Cyclic queries are forbidden')
+
+        return qs.get(slug=args['slug'])
+
+    def resolve_locales(self, args, context, info):
+        qs = LocaleModel.objects.all()
+        fields = get_fields(info)
+
+        if 'locales.localizations' in fields:
+            qs = qs.prefetch_related('project_locale__project')
+
+        if 'locales.localizations.project.localizations' in fields:
+            raise Exception('Cyclic queries are forbidden')
+
+        return qs
+
+    def resolve_locale(self, args, context, info):
+        qs = LocaleModel.objects
+        fields = get_fields(info)
+
+        if 'locale.localizations' in fields:
+            qs = qs.prefetch_related('project_locale__project')
+
+        if 'locale.localizations.project.localizations' in fields:
+            raise Exception('Cyclic queries are forbidden')
+
+        return qs.get(code=args['code'])
```

Comment:
There's a pattern to those 4 ``resolve`` methods. Maybe you can refactor them into a more generic function that would be called with different arguments by each method?

Reply:
I didn't want to refactor too early. I'm leaning towards proposing that we keep these methods as they are for now and try to abstract them when doing research into upgrading to Graphene 2.0.

For instance, we might want to move the cyclic-dependency check to a decorator. `get_fields` could also be a decorator.

Reply:
Same here, sounds good to me!